tests/e2e/sync_test.go

```diff
@@ -0,0 +1,114 @@
+//go:build e2e_tests
+// +build e2e_tests
+
+package e2e
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+var DECK = filepath.Join("../../deck")
+
+func getKongAddress() string {
+	if os.Getenv("KONG_ADDRESS") != "" {
+		return os.Getenv("KONG_ADDRESS")
+	}
+	return "http://localhost:8001"
+}
+
+// cleanUpEnv removes all existing entities from Kong.
+func cleanUpEnv() error {
+	cmd := exec.Command(
+		DECK,
+		"--kong-addr",
+		getKongAddress(),
+		"reset",
+		"--force",
+	) // #nosec G204
+
+	var stderr bytes.Buffer
+	cmd.Stderr = &stderr
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf(stderr.String())
+	}
+	return nil
+}
+
+func normalizeOutput(content string) string {
+	lines := strings.Split(strings.TrimSuffix(content, "\n"), "\n")
+	sort.Strings(lines)
+	for i := range lines {
+		lines[i] = strings.TrimSpace(lines[i])
+	}
+	return strings.Join(lines, "\n")
+}
+
+func loadExpectedOutput(path string) (string, error) {
+	content, err := ioutil.ReadFile(path)
+	if err != nil {
+		return "", err
+	}
+	return string(content), nil
+}
+
+func testDeckOutput(t *testing.T, outputPath string, got string) {
+	expected, err := loadExpectedOutput(outputPath)
+	if err != nil {
+		t.Errorf(err.Error())
+	}
+	expectedNormalized := normalizeOutput(expected)
+	obtainedNormalized := normalizeOutput(got)
+	if !reflect.DeepEqual(obtainedNormalized, expectedNormalized) {
+		t.Errorf(cmp.Diff(obtainedNormalized, expectedNormalized))
+	}
+}
+
+func Test_Sync_output(t *testing.T) {
+	tests := []struct {
+		name string
+	}{
+		{
+			name: "create_service",
+		},
+	}
+
+	if err := cleanUpEnv(); err != nil {
+		t.Errorf(err.Error())
+	}
+
```

Comment:
You are mixing table-driven tests with t.Run() based tests.
The former has most of the data in a table and then the same logic across all t.Run().
Later has no table of data, but each function does the setup and teardown.

I think it makes sense to do table-driven tests here.
The table struct should take in an "input file name", "expected output file name", and then we simply do the same thing across multiple tests.
t.Run() then has the common logic.
You only need a bit of refactoring to get this done, should be quick but the result with be much better (especially as we add more tests).

Reply:
The reason why I was doing this is because I was thinking to have the different test cases to be "incremental" (e.g. create a service, re-create it and expect to see no change, add a route and test the diff etc), so having an initial "setup" to run just once for every test suite seemed the way to go.

If we don't want to mix things, we can have something like "initial state" as part of the table struct (along with "input file name" and "expected output file name") being a kong file to deploy at the beginning of each testcase, and a teardown step at the end of each testcase.

Reply:
It would be best to have small atomic tests. Where we need to compare against an existing state, the "setup" phase for that test should do that setup. This ensures two things:
- A developer can run each and every test, debug it individually.
- Ordering of tests doesn't matter.
- Maintaining state across tests in relation to Kong's database can lead to situations that are annoyingly hard to debug.
- One test failing won't lead to failure of other (dependent) tests.
- Changing one test due to a bug won't require changing any other test.
- Small individual tests decrease the likelihood of tests being buggy (and resulting in code being buggy).

Reply:
Okay I see what I was doing/thinking wrong, makes sense. Thanks for the thoughtful explanation!